tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,223 @@
+    @initialize(key=paths, data=st.binary(min_size=0, max_size=100), target=keys_bundle)
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        self.store.set(key, data_buf)
+        self.model[key] = data_buf
+        note("init key")
+        return key
+
+    @initialize(key=keys_bundle, data=st.data(), target=key_ranges_bundle)
+    def init_key_ranges(self, data, key):
+        note("init key ranges")
+        key_st = st.just(key)
+        key_ranges_st = key_ranges(keys=key_st)
+        return data.draw(key_ranges_st)
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle)
+    def get(self, key) -> None:
+        note(f"(get) {key=}")
+        store_value = self.store.get(key)
+        # to bytes here necessary (on model and store) because data_buf set to model in set()
+        assert self.model[key].to_bytes() == (store_value.to_bytes())
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle, key_range=key_ranges_bundle)
+    def get_partial_values(self, key, key_range) -> None:
+        note(f"(get partial) {key_range=}")
+        obs_maybe = self.store.get_partial_values(key_range)
+        observed = []
+
+        for obs in obs_maybe:
+            assert obs is not None
+            observed.append(obs.to_bytes())
+
+        model_vals_ls = []
+
+        for key, byte_range in key_range:
+            model_vals = self.model[key]
+            start = byte_range[0] or 0
+            step = byte_range[1]
+            stop = start + step if step is not None else None
+            assert model_vals[start:stop] is not None
+            model_vals_ls.append(model_vals[start:stop])
+
+        assert all(
+            obs == exp.to_bytes() for obs, exp in zip(observed, model_vals_ls, strict=True)
+        ), (
+            observed,
+            model_vals_ls,
+        )
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def delete(self, data) -> None:
+        path_st = data.draw(st.sampled_from(sorted(self.model.keys())))
+        note(f"(delete) Deleting {path_st=}")
+
+        self.store.delete(path_st)
+        del self.model[path_st]
+
+    @rule()
+    def clear(self):
+        assert not self.store.mode.readonly
+        note("(clear)")
+        self.store.clear()
+        self.model.clear()
+        # check that model was cleared
+        assert len(self.model.keys()) == 0
+
+    @rule()
+    def empty(self) -> None:
+        note("(empty)")
+
+        # make sure they either both are or both aren't (same state)
+        assert self.store.empty() == (not self.model)
+
+    @rule(key=keys_bundle)
+    def exists(self, key) -> None:
+        #
+        note("(exists)")
+        # make sure same state
+        assert self.store.exists(key) == (key in self.model)
```

Comment:
```suggestion
        assert self.store.exists(key)
        assert key in self.model
```
If you're drawing from `keys_bundle` it should always exist. If it doesn't, you've made an error in the test.

This is good IMO, you're double checking your test.

Reply:
I think this doesn't apply anymore now that we're not using bundles. going to go back to the old test where its asserting that if store.exists() is true, key in self.model also true and vice versa, rather than that both are true